lib/charms/bind/v0/dns_record.py

```diff
@@ -0,0 +1,696 @@
+import itertools
+import json
+import logging
+from enum import Enum
+from typing import Dict, List, Optional
+from uuid import UUID
+
+import ops
+from pydantic import BaseModel, Field, IPvAnyAddress, ValidationError
+
+logger = logging.getLogger(__name__)
+
+DEFAULT_RELATION_NAME = "dns_record"
+
+
+class Status(str, Enum):
+    """Represent the status values.
+
+    Attributes:
+        APPROVED: approved
+        INVALID_CREDENTIALS: invalid_credentials
+        PERMISSION_DENIED: permission_denied
+        CONFLICT: conflict
+        VALIDATION: validation
+        FAILURE: failure
+        UNKNOWN: unknown
+        PENDING: pending
+    """
+
+    APPROVED = "approved"
+    INVALID_CREDENTIALS = "invalid_credentials"
+    PERMISSION_DENIED = "permission_denied"
+    CONFLICT = "conflict"
+    VALIDATION = "validation"
+    FAILURE = "failure"
+    UNKNOWN = "unknown"
+    PENDING = "pending"
+
+
+class RecordType(str, Enum):
+    """Represent the DNS record types.
+
+    Attributes:
+        A: A
+        AAAA: AAAA
+        CNAME: CNAME
+        MX: MX
+        DKIM: DKIM
+        SPF: SPF
+        DMARC: DMARC
+        TXT: TXT
+        CAA: CAA
+        SRV: SRV
+        SVCB: SVCB
+        HTTPS: HTTPS
+        PTR: PTR
+        SOA: SOA
+        NS: NS
+        DS: DS
+        DNSKEY: DNSKEY
+    """
+
+    A = "A"
+    AAAA = "AAAA"
+    CNAME = "CNAME"
+    MX = "MX"
+    DKIM = "DKIM"
+    SPF = "SPF"
+    DMARC = "DMARC"
+    TXT = "TXT"
+    CAA = "CAA"
+    SRV = "SRV"
+    SVCB = "SVCB"
+    HTTPS = "HTTPS"
+    PTR = "PTR"
+    SOA = "SOA"
+    NS = "NS"
+    DS = "DS"
+    DNSKEY = "DNSKEY"
+
+
+class RecordClass(str, Enum):
+    """Represent the DNS record classes.
+
+    Attributes:
+        IN: IN
+    """
+
+    IN = "IN"
+
+
+class DNSProviderData(BaseModel):
+    """Represent the DNS provider data.
+
+    Attributes:
+        uuid: UUID for the domain request.
+        status: status for the domain request.
+        description: status description for the domain request.
+    """
+
+    uuid: UUID
+    status: Status
+    description: Optional[str] = None
+
+    def to_relation_data(self) -> Dict[str, str]:
+        """Convert an instance of DNSProviderData to the relation representation.
+
+        Returns:
+            Dict containing the representation.
+        """
+        result = {"uuid": str(self.uuid), "status": self.status.value}
+        if self.description:
+            result["description"] = self.description
+        return result
+
+    @classmethod
+    def from_relation_data(cls, relation_data: Dict[str, str]) -> "DNSProviderData":
+        """Initialize a new instance of the DNSProviderData class from the relation data.
+
+        Args:
+            relation_data: the relation data.
+
+        Returns:
+            A DNSRecordProviderData instance.
+        """
+        try:
+            status = Status(relation_data.get("status"))
+        except ValueError:
+            status = Status.UNKNOWN
```

Comment:
I think you can use the `model_validate` method from pydantic2 to have a more generic serialization here
See: https://github.com/canonical/digest-squid-auth-helper/blob/first-charm-version/src/charm_state.py#L93

Reply:
Is this comment in the correct place? I'm not sure I follow

Reply:
Sorry, the comment is wrong, I commented too fast on that one.
BUT you should probably leverage the `_missing_` classmethod on enum (see https://docs.python.org/3/library/enum.html#enum.Enum._missing_)
instead defining the default value here